README.md
# rdrugshortages

Access drug shortage data in R. This package currently supports accessing the following databases

1. [Drug Shortages Canada](https://drugshortagescanada.ca)
    Canada-wide mandatory shortage reporting database. You must create an
    account to access the database via the API but you can search and export
    subsets of the data on the site.
    
1. [Drug Product Database](https://www.canada.ca/en/health-canada/services/drugs-health-products/drug-products/drug-product-database.html)
    This database, by Health Canada, provides a current and historical snapshot
    of all available drug products, and includes supplementary information such
    as manufacturer information. There are no shortage reports in this database,
    but it can be used in conjunction with shortage reports by cross-referencing
    the Drug Identification Number (DIN) field.

## Installation Instructions

```r
install.packages('rdrugshortages')

# Or, install the latest development version:
devtools::install_github('pipitone/rdrugshortages')
```

## Use

As mentioned above, you must create an account with Drug Shortages Canada (DSC)
in order to access the database via the API.  Once you have done this, you can
put your credentials in your
[`.Renviron`](https://www.dartistics.com/renviron.html) file which gets loaded
every time you start R:

```
# ~/.Renviron
dsc.email = "youremail@domain.com"
dsc.password = "yourpassword"
```

You can check that this worked by typing the following into a new R session: 

```r
> Sys.getenv('dsc.email')
[1] "youremail@domain.com"
```

You can access the search API via the function `dsc_search()`. This function
will automatically attempt to log in to with your account credentials when you
first use it, and then return the results of the search: 

```r
> results = dsc_search(term="diltiazem")
> head(results)
# A tibble: 6 x 93
     id created_date din   company_name atc_number atc_description discontinuation… updated_date status
  <int> <chr>        <chr> <chr>        <chr>      <chr>           <chr>            <chr>        <chr> 
1 81418 2019-04-11T… 0237… TEVA CANADA… C08DB      SELECTIVE CALC… 2019-03-15T00:0… 2019-04-11T… disco…
2 82552 2019-04-26T… 0237… TEVA CANADA… C08DB      SELECTIVE CALC… NA               2019-04-26T… activ…
3 82555 2019-04-26T… 0237… TEVA CANADA… C08DB      SELECTIVE CALC… NA               2019-04-26T… activ…
4 61340 2018-09-12T… 0237… TEVA CANADA… C08DB01    SELECTIVE CALC… NA               2018-10-11T… activ…
5  9901 2017-05-10T… 0237… ACTAVIS PHA… C08DB      SELECTIVE CALC… NA               2017-09-19T… resol…
6 81381 2019-04-11T… 0237… TEVA CANADA… C08DB      SELECTIVE CALC… 2019-01-09T00:0… 2019-04-11T… disco…
```
The result set returned by `dsc_search()` is a
[`tibble`](https://tibble.tidyverse.org/) which is based on the JSON blob
returned by the API.

The full specification for the DSC API is available [here](https://www.drugshortagescanada.ca/blog/52). 

Note that `rdrugshortages` assembles the entire result set by making multiple
requests to the API when there are more results than can be returned by the API
in a single request. You should be aware that API has been rate-limited to 1000
requests per hour.

To download the entire DSC database, simple run `dsc_search()` without any
search terms.


NEWS.md
# rdrugshortages 0.9
- Initial release that supports the Drug Shortages Canada search API
- Also included are helper functions to download and load the Health Canada Drug
  Product Database extracts


tests/testthat.R
library(testthat)
library(rdrugshortages)

test_check("rdrugshortages")


tests/testthat/test-dpd.R
context("DPD")

test_that("We can download the DPD and load it correctly", {
#  testthat::skip_on_cran()

  data_dir <- tempfile("dpd")
  dpd_download(data_dir)

  expect_gt(length(list.files(data_dir)), 45)
  expect_length(dir(data_dir, pattern="*.txt"), 45)

  dpd <- dpd_load(data_dir)
  expect_is(dpd, 'list')

  tables <- c("comp","form","ingred","package","pharm","drug","route","sched","status","ther","vet")
  for (t in tables) {
    expect_true(t %in% names(dpd))
    expect_is(dpd[[t]], "data.frame")
    expect_true("DRUG_CODE" %in% names(dpd[[t]]))
  }

  unlink(data_dir, recursive=T)
})


tests/testthat/test-dsc_search.R
context("DSC Search")
library(rdrugshortages)

test_that <- function(desc, code) {
    Sys.unsetenv("dsc.email")
    Sys.unsetenv("dsc.password")
    .state$dsc.authtoken <- "token"
    testthat::test_that(desc, code)
}

with_mock_api({
    test_that("Search should attempt to get an authtoken if one is not set", {
        .state$dsc.authtoken <- NULL
        expect_error(dsc_search(report_id = 75200), "*.password not provided")
    })

    test_that("We can get a single report", {
        results <- dsc_search(report_id = 75200)
        expect_is(results, "data.frame")
        expect_equal(nrow(results), 1)
    })

    test_that("We can get a single page of results", {
        results <- dsc_search(max_pages = 1)
        expect_is(results, "data.frame")
        expect_equal(nrow(results), 50)
    })

    test_that("We can get a multiple pages of results", {
        results <- dsc_search(max_pages = 2)
        expect_is(results, "data.frame")
        expect_equal(nrow(results), 100)
    })

    test_that("We can get a single page of json results", {
        results <- dsc_search(format = "json", max_pages = 1)
        expect_is(results, "data.frame")
        expect_equal(nrow(results), 50)
    })

    test_that("An error is thrown if we specify an unknown format", {
        expect_error(dsc_search(format = "csv"), "unknown format")
    })

    test_that("We can return 0 results", {
        results <- dsc_search(din = "NaN")
        expect_equal(nrow(results), 0)
    })

    test_that("All API v1 parameters can be used", {
        results <- dsc_search(max_pages = 1,
                              orderby = "id",
                              order = "asc",
                              filter_status = "resolved",
                              term = "venlafaxine",
                              din = 1,
                              report_id = 1)
        expect_equal(nrow(results), 0)
    })

    test_that("Using non-API v1 parameters causes an error", {
        expect_error(dsc_search(page = 1), "Unexpected search parameter")
        expect_error(dsc_search(total = 1), "Unexpected search parameter")
        expect_error(dsc_search(remaining = 1), "Unexpected search parameter")
        expect_error(dsc_search(data = 1), "Unexpected search parameter")
        expect_error(dsc_search(total_pages = 1), "Unexpected search parameter")
        expect_error(dsc_search(not_a_parameter = 1), "Unexpected search parameter")
    })
})


R/drugproductdatabase.R
#' Downloads the DPD extracts and unzips them
#'
#' This function downloads all of the available sections (current, inactive,
#' etc..) of the current DPD extract.
#'
#' @param data_dir  target directory to download to
#' @export
dpd_download <- function(data_dir = "data/") {
    dpd_base_url <- "https://www.canada.ca/content/dam/hc-sc/documents/services/drug-product-database/"

    if (!dir.exists(data_dir)) {
        dir.create(data_dir)
    }

    for (f in c("allfiles.zip", "allfiles_ia.zip", "allfiles_ap.zip", "allfiles_dr.zip")) {
        destfile <- file.path(data_dir, f)
        url <- paste0(dpd_base_url, f)
        utils::download.file(url, destfile, quiet = T)
        utils::unzip(destfile, exdir = data_dir)
    }
}

#' Loads the DPD into memory from a downloaded extract
#'
#' Each section of the DPD extract is merged together on a per-table basis.
#'
#' @param data_dir    target directory where extract is stored. Use
#'   dpd_download() to fetch and unpack the latest extract.
#' @return A named list of each table as a data.frame
#' @export
#' @importFrom magrittr %>%
dpd_load <- function(data_dir = "data/") {
    tables <- list(
      comp = c("DRUG_CODE", "MFR_CODE", "COMPANY_CODE", "COMPANY_NAME", "COMPANY_TYPE", "ADDRESS_MAILING_FLAG", "ADDRESS_BILLING_FLAG", "ADDRESS_NOTIFICATION_FLAG",  "ADDRESS_OTHER", "SUITE_NUMBER", "STREET_NAME", "CITY_NAME", "PROVINCE", "COUNTRY", "POSTAL_CODE", "POST_OFFICE_BOX", "PROVINCE_F", "COUNTRY_F"),
      form = c("DRUG_CODE", "PHARM_FORM_CODE", "PHARMACEUTICAL_FORM", "PHARMACEUTICAL_FORM_F"),
      ingred = c("DRUG_CODE", "ACTIVE_INGREDIENT_CODE", "INGREDIENT",  "INGREDIENT_SUPPLIED_IND", "STRENGTH", "STRENGTH_UNIT", "STRENGTH_TYPE", "DOSAGE_VALUE", "BASE", "DOSAGE_UNIT", "NOTES", "INGREDIENT_F", "STRENGTH_UNIT_F", "STRENGTH_TYPE_F", "DOSAGE_UNIT_F"),
      package = c("DRUG_CODE", "UPC", "PACKAGE_SIZE_UNIT", "PACKAGE_TYPE", "PACKAGE_SIZE", "PRODUCT_INFORMATION",  "PACKAGE_SIZE_UNIT_F", "PACKAGE_TYPE_F"),
      pharm = c("DRUG_CODE", "PHARMACEUTICAL_STD"),
      drug = c("DRUG_CODE", "PRODUCT_CATEGORIZATION", "CLASS",  "DRUG_IDENTIFICATION_NUMBER", "BRAND_NAME", "DESCRIPTOR", "PEDIATRIC_FLAG", "ACCESSION_NUMBER", "NUMBER_OF_AIS", "LAST_UPDATE_DATE", "AI_GROUP_NO", "CLASS_F", "BRAND_NAME_F", "DESCRIPTOR_F"),
      route = c("DRUG_CODE", "ROUTE_OF_ADMINISTRATION_CODE", "ROUTE_OF_ADMINISTRATION", "ROUTE_OF_ADMINISTRATION_F"),
      sched = c("DRUG_CODE", "SCHEDULE", "SCHEDULE_F"),
      status = c("DRUG_CODE", "CURRENT_STATUS_FLAG", "STATUS", "HISTORY_DATE", "STATUS_F", "LOT_NUMBER",  "EXPIRATION_DATE"),
      ther = c("DRUG_CODE", "TC_ATC_NUMBER", "TC_ATC", "TC_AHFS_NUMBER", "TC_AHFS", "TC_ATC_F", "TC_AHFS_F"),
      vet = c("DRUG_CODE",  "VET_SPECIES", "VET_SUB_SPECIES", "VET_SPECIES_F"))

    dpd <- list()
    for (table_name in names(tables)) {
        headers <- tables[[table_name]]

        dpd[[table_name]] <- dir(data_dir, pattern = paste0(table_name, ".*.txt$"), full.names = T) %>%
          purrr::keep(function(x) file.size(x) > 0) %>%
          purrr::map(readr::read_csv,
            col_names = headers,
            locale = readr::locale("en",  date_format = "%d-%b-%Y"),
            col_types = purrr::map(headers, function(x) dplyr::case_when(endsWith(x, "_DATE") ~ "D", x == "DRUG_CODE" ~ "i", T ~ "c")) %>%
              stats::setNames(headers), progress = F) %>%
          purrr::reduce(rbind)
    }

    return(dpd)
}


R/drugshortagescanada.R
#' Login to the drugshortagescanada.ca web API
#'
#' This function logs in to the DSC web API and stores the temporary
#' authentication token so that it can be used by other API calls.
#'
#' You shouldn't usually need to call this function directly, as dsc_search()
#' will automatically do so.
#'
#' The username and password can be passed into the function explicitly, or can
#' be set via environment variables in .Renviron, or directly like so:
#'
#'   Sys.setenv('dsc.email' = 'bill.gates@microsoft.com')
#'   Sys.setenv('dsc.password' = 'passw0rd!')
#'
#'   dsc_login()
#'
#' @param email Your email address used to log into the DSC
#' @param password Your password
#' @param reuse_authtoken If an authentication token has already been retrieved
#'   this session, then return it without logging in again via the API
#' @return authentication token
#' @export
dsc_authtoken <- function(email, password, reuse_authtoken = F) {
    if (reuse_authtoken & !is.null(.state$dsc.authtoken)) {
        return(.state$dsc.authtoken)
    }
    if (missing(email)) {
        email <- Sys.getenv("dsc.email", unset = NA)
    }
    if (missing(password)) {
        password <- Sys.getenv("dsc.password", unset = NA)
    }
    if (is.na(email) | is.na(password)) {
        stop("email and/or password not provided.")
    }
    r <- httr::POST("https://www.drugshortagescanada.ca/api/v1/login", body = list(email = email, password = password))
    httr::stop_for_status(r, httr::content(r, as = "text"))
    authtoken <- httr::headers(r)$`auth-token`
    if (is.null(authtoken)) {
        stop("auth-token not provided by endpoint")
    }
    .state$dsc.authtoken <- authtoken

    authtoken
}

#' A single page of search results
#'
#' @param ... name = value pairs of query parameters.
#' @return An R representation of JSON search results via jsonlite::fromJSON
#'   records.
.dsc_search_once <- function(...) {
    authtoken <- dsc_authtoken(reuse_authtoken = T)
    r <- httr::GET("https://www.drugshortagescanada.ca/api/v1/search", httr::add_headers(`auth-token` = authtoken), query = list(...))
    httr::stop_for_status(r, httr::content(r, as = "text"))
    return(jsonlite::fromJSON(httr::content(r, as = "text", encoding = "UTF-8")))
}

#' Return all records from a search
#'
#' @param ... name = value pairs of query parameters.
#' @param max_pages Stop after this many pages of results
#'
#' @return  nested data.frame of results
.dsc_search_all <- function(..., max_pages = Inf) {
    query <- list(...)

    if (is.null(query$limit)) {
        query$limit <- 1000  # some optimistically large number. 0 doesn't work.
    }

    results <- do.call(.dsc_search_once, query)
    all_results <- list(results$data)
    pages <- 1

    while (results$remaining > 0 & pages < max_pages) {
        query$offset <- results$offset + nrow(results$data)
        results <- do.call(.dsc_search_once, query)
        all_results <- c(all_results, list(results$data))
        pages <- pages + 1
    }

    if (!length(unlist(all_results))) {
        return(as.data.frame(all_results))
    } else {
        return(jsonlite::rbind_pages(all_results))
    }
}

#' Query the DSC
#'
#' The search API is documented here:
#' https://www.drugshortagescanada.ca/blog/52
#'
#'
#'
#' @param ... A named list of query parameters.
#'   See the DSC Web API documentation for details on query parameters:
#'   https://www.drugshortagescanada.ca/blog/52
#' @param format The default 'tidy' returns a tibble with columns converted to
#'   the appropriate data types. Use 'json' to get results as returned by
#'   the API in JSON format after using jsonlite::rbind_pages to concatenate
#'   multiple pages of results.
#' @param max_pages Limit the number of requests made to the API. The number of
#'   results returned will depend on the total results, and the limit and offset
#'   parameters if supplied.
#' @return Results formatted according to format
#' @export
dsc_search <- function(..., format = "tidy", max_pages = Inf) {
    v1_params <- c("limit", "offset", "orderby", "order", "filter_status", "term", "din", "report_id")
    query <- list(...)
    unknown_params <- setdiff(names(query), v1_params)
    if (length(unknown_params) > 0) {
        unknown <- paste(unknown_params, collapse = ", ")
        allowed <- paste(v1_params, collapse = ", ")
        stop(paste0("Unexpected search parameter(s): ", unknown, "\nAllowed parameters are: ", allowed))
    }

    if (!(format %in% c("tidy", "json"))) {
        stop("unknown format")
    }

    results <- do.call(.dsc_search_all, list(..., max_pages = max_pages))
    return(switch(format, tidy = .dsc_tidy_results(results), json = results))
}

#' Tidy up DSC search results
#'
#' Accepts the raw JSON-to-data.frame search results object and converts columns
#' to their appropriate data types (dates, factors, etc)
#'
#' @param results search results as data.frame
#' @return tibble
.dsc_tidy_results <- function(results) {
    dplyr::as_tibble(jsonlite::flatten(results))
}
